### PR TITLE
Task2

### PR DESCRIPTION
**Pull request by martinmilev**
Ready for review. Tests included.
[presentation](https://martinmilev.github.io/htmlPages/servlets/task2.html)

### CODE REVIEW

**Comment by freekman** [comment]
Every PR is see one thing .idea boilerplate code is it so hard to remove it ?

**Comment by freekman** [comment]
is this PR ready for review

**Comment by martinmilev** [comment]
This PR is ready for review.

**Comment by KrasimirRaikov** [comment]
Do you really need all this xml-s, from .idea?

File: src/com/clouway/servlets/pages/Demo.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ */
+public class Demo {
+  public static void main(String[] args) throws IOException {
+    Map<String, String> pages = new LinkedHashMap<String, String>() {{
+      put("index", Files.toString(new File("src/com/clouway/servlets/pages/resources/index.html"), Charsets.UTF_8));
+      put("page1", Files.toString(new File("src/com/clouway/servlets/pages/resources/page1.html"), Charsets.UTF_8));
+      put("page2", Files.toString(new File("src/com/clouway/servlets/pages/resources/page2.html"), Charsets.UTF_8));
+      put("page3", Files.toString(new File("src/com/clouway/servlets/pages/resources/page3.html"), Charsets.UTF_8));
+    }};
+    PageHandler handler = new PageHandler(pages);
```
Comment (freekman) [inline src/com/clouway/servlets/pages/Demo.java]: I think that this class PageHandler can be removed

File: src/com/clouway/servlets/pages/adapter/PageHandler.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  private final Map<String, String> pages;
+
+  public PageHandler(Map<String, String> pages) {
+    this.pages = pages;
+  }
+
+  @Override
+  public String getPage(String link) {
+    if (link != null) {
+      if ("link1".equals(link)) {
```
Comment (freekman) [inline src/com/clouway/servlets/pages/adapter/PageHandler.java]: extreme duplication

File: test/com/clouway/servlets/pages/adapter/PageHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import static org.hamcrest.MatcherAssert.assertThat;
+
+/**
+ * @author Martin Milev <martinmariusmilev@gmail.com>
+ */
+public class PageHandlerTest {
+  @Test
+  public void happyPath() throws Exception {
+    Map<String, String> pages = new LinkedHashMap<String, String>() {{
+      put("index", Files.toString(new File("src/com/clouway/servlets/pages/resources/index.html"), Charsets.UTF_8));
```
Comment (freekman) [inline test/com/clouway/servlets/pages/adapter/PageHandlerTest.java]: why are you reading from a file to get the content of a page?

File: test/com/clouway/servlets/pages/adapter/PageHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+/**
+ * @author Martin Milev <martinmariusmilev@gmail.com>
+ */
+public class PageHandlerTest {
+  @Test
+  public void happyPath() throws Exception {
```
Comment (freekman) [inline test/com/clouway/servlets/pages/adapter/PageHandlerTest.java]: missing test cases

File: src/com/clouway/servlets/pages/Demo.java (hunk 1 above)
Comment (martinmilev) [inline src/com/clouway/servlets/pages/Demo.java]: fixed

File: test/com/clouway/servlets/pages/adapter/PageHandlerTest.java (hunk 4 above)
Comment (martinmilev) [inline test/com/clouway/servlets/pages/adapter/PageHandlerTest.java]: PageHandler removed

File: test/com/clouway/servlets/pages/adapter/PageHandlerTest.java (hunk 3 above)
Comment (martinmilev) [inline test/com/clouway/servlets/pages/adapter/PageHandlerTest.java]: PageHandler removed

File: src/com/clouway/servlets/pages/Demo.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+      put("link3", Files.toString(new File("src/com/clouway/servlets/pages/resources/page3.html"), Charsets.UTF_8));
+    }};
+    HtmlTemplate template = new HtmlTemplate();
+
+    Map<String, HttpServlet> servlets = new LinkedHashMap<String, HttpServlet>() {{
+      put("/", new PagesServlet(template, pages));
+    }};
+
+    JettyServer server = new JettyServer(8080, servlets);
+    server.start();
```
Comment (freekman) [inline src/com/clouway/servlets/pages/Demo.java]: add ```server.join()```

File: test/com/clouway/servlets/pages/FakeHttpServletRequest.java [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+import java.util.Collection;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+
+/**
+ * @author Martin Milev <martinmariusmilev@gmail.com>
+ */
+public class FakeHttpServletRequest implements HttpServletRequest {
```
Comment (freekman) [inline test/com/clouway/servlets/pages/FakeHttpServletRequest.java]: you have merged Your first PR why am still seeing these fakes

File: test/com/clouway/servlets/pages/adapter/HtmlTemplateTest.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+import org.junit.Test;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+/**
+ * @author Martin Milev <martinmariusmilev@gmail.com>
+ */
+public class HtmlTemplateTest {
```
Comment (freekman) [inline test/com/clouway/servlets/pages/adapter/HtmlTemplateTest.java]: .

File: src/com/clouway/servlets/pages/adapter/HtmlTemplate.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
+package com.clouway.servlets.pages.adapter;
+
+import com.clouway.servlets.pages.core.Template;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+public class HtmlTemplate implements Template {
```
Comment (freekman) [inline src/com/clouway/servlets/pages/adapter/HtmlTemplate.java]: .

File: src/com/clouway/servlets/pages/servlets/PagesServlet.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+  }
+
+  private void setMsg(String link) throws IOException {
+    if (link != null) {
+      if (link.contains("link")) {
+        template.setTemplateValue(pages.get(link));
+      }
+      if (link.contains("page")) {
+        template.put("page", "Last visited: " + link + ".");
+        template.setTemplateValue(pages.get("index"));
```
Comment (freekman) [inline src/com/clouway/servlets/pages/servlets/PagesServlet.java]: can you extract the ```"index"``` in a variable for better readability

File: test/com/clouway/servlets/pages/FakeHttpServletRequest.java (hunk 6 above)
Comment (martinmilev) [inline test/com/clouway/servlets/pages/FakeHttpServletRequest.java]: I don't need those files in the pr?

Comment (freekman) [inline test/com/clouway/servlets/pages/FakeHttpServletRequest.java]: so we cant reuse the FakeRequest and FakeResponse

**Pull request closed without merging**